# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/security/trusted-types/TrustedTypePolicyFactory.cpp
# Commit: 4e3e9ade5089
# Full Hash: 4e3e9ade50894ab253b5e02423e59a4d0df1503c
# Author: Mirko Brodesser <mbrodesser@igalia.com>
# Date: 2024-06-19 21:39:42
# Regressor Bug: 1901491
# File Overlap Count: 1
# Description:
#   Bug 1901491: part 3) Implement `ShouldTrustedTypePolicyCreationBeBlockedByCSP` for Windows (not Workers) including reporting exceptions, excluding reporting violations. r=tschuster,webidl,saschanaz
#   
#   Parsing the trusted-types directive's values is not added to
#   `nsCSPParser::sourceList` because the latter supports keywords not
#   supported by the trusted-types directive. E.g. 'report-sample'.
# ==============================================================================

diff -r dd03de7652a1 -r 4e3e9ade5089 dom/security/trusted-types/TrustedTypePolicyFactory.cpp
--- a/dom/security/trusted-types/TrustedTypePolicyFactory.cpp	Wed Jun 19 12:51:28 2024 +0000
+++ b/dom/security/trusted-types/TrustedTypePolicyFactory.cpp	Wed Jun 19 12:51:29 2024 +0000
@@ -7,8 +7,10 @@
 #include "mozilla/dom/TrustedTypePolicyFactory.h"
 
 #include "mozilla/AlreadyAddRefed.h"
+#include "mozilla/ErrorResult.h"
 #include "mozilla/RefPtr.h"
 #include "mozilla/dom/TrustedTypePolicy.h"
+#include "mozilla/dom/nsCSPUtils.h"
 
 namespace mozilla::dom {
 
@@ -19,10 +21,52 @@
   return TrustedTypePolicyFactory_Binding::Wrap(aCx, this, aGivenProto);
 }
 
+static constexpr auto kWildcard = u"*"_ns;
+
+bool TrustedTypePolicyFactory::ShouldTrustedTypePolicyCreationBeBlockedByCSP(
+    const nsAString& aPolicyName) const {
+  // CSP-support for Workers will be added in
+  // <https://bugzilla.mozilla.org/show_bug.cgi?id=1901492>.
+  // That is, currently only Windows are supported.
+  nsIContentSecurityPolicy* csp = mGlobalObject->GetAsInnerWindow()->GetCsp();
+
+  if (csp) {
+    uint32_t numPolicies = 0;
+    csp->GetPolicyCount(&numPolicies);
+
+    for (uint64_t i = 0; i < numPolicies; ++i) {
+      const nsCSPPolicy* policy = csp->GetPolicy(i);
+      if (policy->hasDirective(
+              nsIContentSecurityPolicy::TRUSTED_TYPES_DIRECTIVE)) {
+        if (policy->ShouldCreateViolationForNewTrustedTypesPolicy(
+                aPolicyName, mCreatedPolicyNames)) {
+          // TODO: create violation, populate it with data and report it. See
+          // <https://bugzilla.mozilla.org/show_bug.cgi?id=1901510>.
+
+          if (policy->getDisposition() == nsCSPPolicy::Disposition::Enforce) {
+            return true;
+          }
+        }
+      }
+    }
+  }
+
+  return false;
+}
+
 already_AddRefed<TrustedTypePolicy> TrustedTypePolicyFactory::CreatePolicy(
     const nsAString& aPolicyName,
-    const TrustedTypePolicyOptions& aPolicyOptions) {
-  // TODO: add CSP support.
+    const TrustedTypePolicyOptions& aPolicyOptions, ErrorResult& aRv) {
+  if (ShouldTrustedTypePolicyCreationBeBlockedByCSP(aPolicyName)) {
+    nsCString errorMessage =
+        "Content-Security-Policy blocked creating policy named '"_ns +
+        NS_ConvertUTF16toUTF8(aPolicyName) + "'"_ns;
+
+    // TODO: perhaps throw different TypeError messages,
+    //       https://github.com/w3c/trusted-types/issues/511.
+    aRv.ThrowTypeError(errorMessage);
+    return nullptr;
+  }
 
   // TODO: add default policy support; this requires accessing the default
   //       policy on the C++ side, hence already now ref-counting policy